File: Go-Beacon/pkg/core/outbox.go
package core

import "sync"

const (
	outboxMaxPackets = 1024             // 队列最大包数
	outboxMaxBytes   = 64 * 1024 * 1024 // 队列最大字节数（64MB）
)

// OutboxNode 是 Outbox 链表中的一个节点，持有一个待发送的数据包。
type OutboxNode struct {
	packet []byte
	next   *OutboxNode
}

// Outbox 是线程安全的单链表队列，用于缓存待发送的结果包。
// 主循环每次心跳时 Drain 全部节点，加密后逐个发送；失败的节点通过 PushFrontList 回退。
type Outbox struct {
	mu    sync.Mutex
	head  *OutboxNode
	tail  *OutboxNode
	count int
	bytes int
}

// Init 初始化 Outbox 队列，清空所有节点和计数器。
func (o *Outbox) Init() {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.head = nil
	o.tail = nil
	o.count = 0
	o.bytes = 0
}

// Free 释放 Outbox 中的所有节点并重置队列。
func (o *Outbox) Free() {
	o.mu.Lock()
	defer o.mu.Unlock()
	o.head = nil
	o.tail = nil
	o.count = 0
	o.bytes = 0
}

// Enqueue 将一个数据包追加到队列尾部。
// 超出容量限制（包数或字节数）时静默丢弃。
func (o *Outbox) Enqueue(packet []byte) {
	if len(packet) == 0 {
		return
	}

	n := &OutboxNode{packet: packet}

	o.mu.Lock()
	defer o.mu.Unlock()
	if o.count >= outboxMaxPackets ||
		len(packet) > outboxMaxBytes ||
		o.bytes > outboxMaxBytes-len(packet) {
		return
	}

	if o.tail != nil {
		o.tail.next = n
	} else {
		o.head = n
	}
	o.tail = n
	o.count++
	o.bytes += len(packet)
}

// Drain 取出队列中的所有节点并清空队列，返回链表头指针。
func (o *Outbox) Drain() *OutboxNode {
	o.mu.Lock()
	defer o.mu.Unlock()

	list := o.head
	o.head = nil
	o.tail = nil
	o.count = 0
	o.bytes = 0
	return list
}

// PushFrontList 将一条链表整体插入队列头部（用于发送失败时回退未发送的包）。
func (o *Outbox) PushFrontList(list *OutboxNode) {
	if list == nil {
		return
	}

	tail := list
	count := 1
	bytes := len(tail.packet)
	for tail.next != nil {
		tail = tail.next
		count++
		bytes += len(tail.packet)
	}

	o.mu.Lock()
	defer o.mu.Unlock()
	tail.next = o.head
	o.head = list
	if o.tail == nil {
		o.tail = tail
	}
	o.count += count
	o.bytes += bytes
}

File: C-Beacon/src/core/outbox.c
#include "beacon_outbox.h"

#ifdef BEACON_TEST
#include "beacon_test_hooks.h"
#endif

#define OUTBOX_MAX_PACKETS 1024
#define OUTBOX_MAX_BYTES   (64 * 1024 * 1024)

/*
 * OutboxInit - 初始化发件箱临界区并清空队列。
 */
VOID OutboxInit(Outbox* outbox)
{
    InitializeCriticalSection(&outbox->lock);
    outbox->head = NULL;
    outbox->tail = NULL;
    outbox->count = 0;
    outbox->bytes = 0;
}

/*
 * OutboxFreeNode - 释放单个发件箱节点的内存。
 */
VOID OutboxFreeNode(OutboxNode* node)
{
    if (!node) return;

    BbFree(&node->packet);
    HeapFree(GetProcessHeap(), 0, node);
}

/*
 * OutboxFree - 排空并释放所有排队的数据包，然后销毁锁。
 */
VOID OutboxFree(Outbox* outbox)
{
    OutboxNode* n = outbox->head;

    /* 遍历链表并释放每个节点 */
    while (n) {
        OutboxNode* next = n->next;
        OutboxFreeNode(n);
        n = next;
    }

    outbox->head = NULL;
    outbox->tail = NULL;
    outbox->count = 0;
    outbox->bytes = 0;
    DeleteCriticalSection(&outbox->lock);
}

/*
 * OutboxEnqueue - 将数据包追加到发件箱队列尾部。
 *   超过上限时丢弃当前数据包；分配失败时也释放。
 */
VOID OutboxEnqueue(Outbox* outbox, ByteBuf packet)
{
    OutboxNode* n = (OutboxNode*)HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(OutboxNode));

    /* 分配失败：丢弃数据包以避免内存泄漏 */
    if (!n) {
        BbFree(&packet);
        return;
    }

    n->packet = packet;

    EnterCriticalSection(&outbox->lock);
    if (outbox->count >= OUTBOX_MAX_PACKETS ||
        packet.len > OUTBOX_MAX_BYTES ||
        outbox->bytes > OUTBOX_MAX_BYTES - packet.len) {
        LeaveCriticalSection(&outbox->lock);
        OutboxFreeNode(n);
        return;
    }

    /* O(1) 追加到队尾 */
    if (outbox->tail) {
        outbox->tail->next = n;
    } else {
        outbox->head = n;
    }
    outbox->tail = n;
    ++outbox->count;
    outbox->bytes += packet.len;
#ifdef BEACON_TEST
    BeaconTestRecord(BEACON_TEST_EVENT_OUTBOX_ENQUEUE, 0, (ULONG)packet.len);
#endif
    LeaveCriticalSection(&outbox->lock);
}

/*
 * OutboxDrain - 原子地分离并返回整个队列。
 *   调用后发件箱为空；调用方拥有所有返回的节点。
 */
OutboxNode* OutboxDrain(Outbox* outbox)
{
    OutboxNode* list;

    EnterCriticalSection(&outbox->lock);
    list = outbox->head;
    outbox->head = NULL;
    outbox->tail = NULL;
    outbox->count = 0;
    outbox->bytes = 0;
    LeaveCriticalSection(&outbox->lock);

    return list;
}

/*
 * OutboxPushFrontList - 将先前排空的列表前置到队列前端。
 *   遍历在锁外完成（传入列表归调用方所有）。
 *   回塞不受 OUTBOX_MAX_* 限制，保证已排队结果不丢失。
 */
VOID OutboxPushFrontList(Outbox* outbox, OutboxNode* list)
{
    OutboxNode* tail;
    SIZE_T count = 1;
    SIZE_T bytes;

    if (!list) return;

    /* 锁外遍历：list 归调用方所有，无需持锁 */
    tail = list;
    bytes = tail->packet.len;
    while (tail->next) {
        tail = tail->next;
        ++count;
        bytes += tail->packet.len;
    }

    /* 拼接到队首 */
    EnterCriticalSection(&outbox->lock);
    tail->next = outbox->head;
    outbox->head = list;
    if (!outbox->tail) {
        outbox->tail = tail;
    }
    outbox->count += count;
    outbox->bytes += bytes;
    LeaveCriticalSection(&outbox->lock);
}
